### PR TITLE
Implementing xrayTraceIDRatioBasedsampler

### PR DESCRIPTION
**Pull request by alanprot**
Signed-off-by: Alan Protasio <approtas@amazon.com>

Hi..

The default Otel TraceIDRatioBased does not work with XRay as it make the decision based on the first 8 bytes of the traceId  -> which are not random on traces generated with the `xray.NewIDGenerator()`.

This PR creates a similar sampler that works on XRAY as it makes the decision based on the last 8 bytes of the trace id.

### CODE REVIEW

File: samplers/aws/xray/go.mod [hunk 1]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/go-logr/logr v1.2.3
 	github.com/go-logr/stdr v1.2.2
 	github.com/stretchr/testify v1.8.1
+	go.opentelemetry.io/contrib/propagators/aws v1.12.0
```
Comment (Aneurysm9) [inline samplers/aws/xray/go.mod:9]: Please ensure `make crosslink` is run to add replace statements for modules in this repo.

File: samplers/aws/xray/traceid_ratio_sampler.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"go.opentelemetry.io/otel/trace"
+)
+
+type traceIDRatioBasedSampler struct {
+	sdktrace.Sampler
+	max  uint64
+	desc string
+}
+
+// NewTraceIDRatioBased creates a sampler based on random number.
```
Comment (Aneurysm9) [inline samplers/aws/xray/traceid_ratio_sampler.go:32]:
```suggestion
// NewTraceIDRatioBased creates a sampler based on (presumably) random parts of the TraceID.
```

**Review by Aneurysm9 — COMMENTED** [review]
Just a couple nits.

File: samplers/aws/xray/go.mod (hunk 1 above)
Comment (alanprot) [reply samplers/aws/xray/go.mod:9]: done

**Review by Aneurysm9 — APPROVED** [review]

**Comment by willarmiros** [comment]
This is interesting to see, because I know in Java the TraceIdRatioBased sampler reads from the trailing bytes rather than the leading bytes by default. We even [have a test for it](https://github.com/aws-observability/aws-otel-java-instrumentation/blob/a011b8cc29ee32b7f668c04ccfdf64cd30de467c/awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsTracerConfigurerTest.java#L60-L84). So it makes me wonder, is the correct fix to make a new sampler for X-Ray IDs only (which may be very confusing for customers to use), or could we change the default traceIdRatioBased sampler to just read from the trailing bytes in the main Go SDK?

**Comment by alanprot** [comment]
> Thanks for creating this PR! This is interesting to see, because I know in Java the TraceIdRatioBased sampler reads from the trailing bytes rather than the leading bytes by default. We even [have a test for it](https://github.com/aws-observability/aws-otel-java-instrumentation/blob/a011b8cc29ee32b7f668c04ccfdf64cd30de467c/awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsTracerConfigurerTest.java#L60-L84). So it makes me wonder, is the correct fix to make a new sampler for X-Ray IDs only (which may be very confusing for customers to use), or could we change the default traceIdRatioBased sampler to just read from the trailing bytes in the main Go SDK?

I had the same question but I guess I took the safer route.. idk if changing the default one will break some other use cases ..

What the others think ?

**Comment by Aneurysm9** [comment]
I think it would be ideal to fix in the base sampler.  I'm not aware of anything that should break if the position of the sampled bits changes.  https://github.com/open-telemetry/opentelemetry-go/pull/3557 may replace this PR.

File: samplers/aws/xray/traceid_ratio_sampler.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		return sdktrace.NeverSample()
+	}
+
+	return &traceIDRatioBasedSampler{
+		desc: fmt.Sprintf("xrayTraceIDRatioBasedSampler{%v}", fraction),
+		max:  uint64(fraction * math.MaxUint64),
+	}
+}
+
+func (s *traceIDRatioBasedSampler) ShouldSample(p sdktrace.SamplingParameters) sdktrace.SamplingResult {
```
Comment (hanyuancheung) [inline samplers/aws/xray/traceid_ratio_sampler.go:49]: Please add a note for exported func.

**Review by hanyuancheung — APPROVED** [review]

**Comment by tomwans** [comment]
Take a look at a related PR: https://github.com/open-telemetry/opentelemetry-go/pull/3557 which brings the Go SDK in line with the Java and Python SDKs, which all by default sample from the trailing bytes.

**Pull request closed without merging**